### ISSUE TITLE
RGB order?

### ISSUE THREAD

**Issue by lichengunc** [karpathy/neuraltalk2#23]
I'm wondering if there is RGB->BGR conversion in the code? 
It looks to me the imgs only goes cropping and RGB mean subtraction in the prepro.py.
Then the imgs are fed into VGGNet.

I guess this won't make too much difference on the results, but is it better to go through normal workflow, i.e., adding BGR conversion?

Thanks,
-Licheng

**Comment by karpathy** [comment]
everything is done in RGB. Why go into BGR at all?

**Comment by lichengunc** [comment]
My reason is we are loading the pretrained model from VGGNet, whose default forwarded option is BGR order, at least for the classification task. I mean there might be of little difference in the end after finetuning CNN side, but would it be better to follow VGGNet workflow in the early feature extraction stage?

**Comment by karpathy** [comment]
doh, wait, I slipped on this. I made the preprocessing be RGB, made sure to swap RB in the mean subtraction and made a mental note to swap the channels in the first layer filters when loading the VGGNet, and then forgot to do it. I'll fix this - either adjusting RGB->BGR in prepro/mean, or fixing the 1st layer conv filters as I originally intended (I think this should be easier).

**Comment by lichengunc** [comment]
Yes, that's what I mean. :-) this may also change your released model - the pretrained checkpoint. Thanks for fixing this.

Just as a side question, I'm a bit confused by the two-stage training:
stage A - fix the CNN-part and train LSTM
stage B - jointly finetune CNN and LSTM. 
However, there are dropout layers in the CNN. I guess the input to LSTM could be unstable due to random drop in stage A. Would it be better to make cnn:evaluate() mode during stage A, and then change it back to cnn:train() in stage B?

**Comment by karpathy** [comment]
I think the pretrained model is fine because it was finetuned.

Good point about dropout... I'm inclined to think that this is probably not a big problem, at least not as big as the RGB BGR issue.

**Comment by lichengunc** [comment]
Yes, I agree. Thank you for the reply.
As for the pretrained model, I mean after you fixed the code, your current model won't fit it. A new one might be released in the future. :-) I will also train it myself. Hope this could make some improvement, like speeding up convergence or higher BLEU.

**Comment by karpathy** [comment]
I'm quite sure it would speed up convergence. Also, I think that changing the first layer filters when loading the caffe model should be backwards compatible to previous checkpoint. What wouldn't be backwards compatible is changing the processing to use BGR images. I'll look into this sometime today hopefully.

**Comment by lichengunc** [comment]
Thanks a lot. I appreciate you share with us this code.
